roles/StackStorm.st2smoketests/tasks/st2chatops.yml

```diff
@@ -1,15 +1,41 @@
 ---
+- name: Prepare temp file to track retry attempts
+  tempfile:
+    state: file
+  register: attempts_count
+  changed_when: no  # skip idempotence check
+
 - name: Verify st2chatops using bin/hubot
   # when editing, make sure it works for at least 2 adapters: 'shell' and 'slack'
-  shell: set -o pipefail && timeout 10 bash -c '(sleep 5; echo exit ) | bin/hubot'
+  shell: >
+    set -o pipefail &&
+    export n=0$(cat $attempts_count_file) &&
+    timeout $(( 9 + 6 ** n )) bash -c '(sleep $(( 4 + 6 ** n )); echo exit ) | bin/hubot';
+    echo $((n+1)) > $attempts_count_file
+  # $a,$n = homegrown exponential backoff (increasing sleep lets the slack client lib retry automatically)
   args:
     chdir: /opt/stackstorm/chatops/
     executable: /bin/bash
   environment:
+    attempts_count_file: "{{ attempts_count.path }}"
     HUBOT_LOG_LEVEL: debug
   register: hubot_output
   failed_when: no
   changed_when: no
+  # any EADDRINUSE error in stdout is normal (an artifact of the js express http framework)
+  # On Travis, with failures when running many builds in parallel, the real error message is:
+  # "Rate limited, will retry # seconds". This comes from @slack/client (used by hubot-slack).
+  # On success (with Slack), the log says: 'SlackBot#authenticated() Found self in RTM start data'
+  until: "'Rate limited, will retry' not in hubot_output.stdout_lines | last"
+  retries: 3
+  # random delay so that retries accross parallel runs don't perfectly align
+  delay: "{{ 15 + (7 | random) }}"
```

Comment:
probably a expression for randomness would be: `{{ range(0, 30) | random }}` or something along those lines (pick any min/max you like) 

if i look at the way that `random()` is implemented it generates a random number and picks one from the sequence (list) that is passed in. Since you're always passing in 7, it is always going to pick the number 7 and add 15.

Reply:
https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters.html#random-number-filter
> To get a random number between 0 and a specified number:
> ```
> "{{ 60 | random }} * * * * root /script/from/cron"
> # => '21 * * * * root /script/from/cron'
> ```

It's an ansible-specific extension on top of the default jinja filter.
https://github.com/ansible/ansible/blob/v2.9.4/lib/ansible/plugins/filter/core.py#L226-L231